### ISSUE TITLE
Question: Implementation of Philips Hue bulbs?

### ISSUE THREAD

**Issue by Jayky** [eliotstocker/Light-Controller#29]
I found this App via F-droid and was really glad to see that even open source apps exist to control wifi bulbs. Unfortunatly I have none of the supported bulbs. So I would like to ask if there will be a connectivity to Hue bulps of Philips? 
Anyway, great work. Hope you will keep on working on this app! 
Maybe you will be able to create an alternative to the Apple app Ambify. Would be great :-)

**Comment by eliotstocker** [comment]
If someone links me to some api documentation ill take a look at it and evaluate what might need to be done (not promising anything) of course without having any of the bulbs the testing will have to be done by someone with the bulbs

**Comment by Jayky** [comment]
Dear Eliot, thanks for your answer. I checked it and found a documentation here: http://www.developers.meethue.com/philips-hue-api
Unfortunatly you need to register there. If this should be a problem, I will do it and send you the ducumentation via email. 
Of course, I know you can't promise anything, but if you would try it, I will try to help you by doing the testing.

**Comment by eliotstocker** [comment]
I'll start looking into this once 1.7 release is out of the way. Thanks for the info.

**Comment by xwbfozuh** [comment]
Perhaps this project would be relevant? https://github.com/thht/ambilike

**Comment by badrange** [comment]
Looks like this is the only open source phone app available - at least according to a quick check on f-droid. I just got my hands on a Philips HUE set and was pretty turned off by them requiring access to media, camera & location for their own app to even open. And you should read their privacy policy, scary! 

There are many out here that would be massively grateful if Light Controller would add Philips HUE support.

**Comment by dudeofawesome** [comment]
I just got a LifX Color 1000 bulb, and am thinking about working on adding support to Light Controller for LifX. It probably wouldn't be too much more work to add Hue too.

**Comment by eliotstocker** [comment]
ok cool, i think the best course for this sort of stuff is to have a plugin system that would allow separate packages to provide a means to control other types of bulbs, ill get some framework together soon :)

**Comment by dudeofawesome** [comment]
Here's some reference info I put together

### Table of features supported by different bulbs

|                  | MiLight RGBW | MiLight RGBWW | MiLight Dual White | LifX + | LifX Color 1000 | LifX White 800 | LifX Z | Hue White and Color | Hue White Ambiance | Hue White | Hue Lightstrip Plus | Hue Lightstrip | Hue Go | Hue Bloom |
| ---------------: | :----------: | :-----------: | :----------------: | :----: | :-------------: | :------------: | :----: | :-----------------: | :----------------: | :-------: | :------------: | :-----------------: | :----: | :-------: |
|         On / Off | ✓ | ✓                                                                | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|       True White | ✓ | ✓                                                                | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✗                                | ✗ | ✗                                | ✗                                |
|              RGB | ✓ | ✓                                                                | ✗ | ✓ | ✓ | ✗ | ✓ | ✓ | ✗ | ✗ | ✓                                | ✓ | ✓                                | ✓                                |
|   RGB Saturation | ✗ | ✗                                                                | ✗ | ✓ | ✓ | ✗ | ✓ | ✓ | ✗ | ✗ | ✓                                | ✓ | ✓                                | ✓                                |
|      Temperature | ✗ | ✓                                                                | ✓ | ✓ | ✓ | ✓ | ✗ | ✓ | ✓ | ✗ | <a href="" title="Sort of">✓</a> | ✗ | <a href="" title="Sort of">✓</a> | <a href="" title="Sort of">✓</a> |
|         Infrared | ✗ | ✗                                                                | ✗ | ✓ | ✗ | ✗ | ✗ | ✗ | ✗ | ✗ | ✗                                | ✗ | ✗                                | ✗                                |
| Internet Control | ? | <a href="" title="Requires V6 bridge, which isn't out yet">?</a> | ? | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|         REST API | ✗ | <a href="" title="Requires V6 bridge, which isn't out yet">✗</a> | ✗ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|     Bare UDP API | ✓ | <a href="" title="Requires V6 bridge, which isn't out yet">✓</a> | ✓ | ✓ | ✓ | ✓ | ✓ | ✗ | ✗ | ✗ | ✗                                | ✗ | ✗                                | ✗                                |
|       Scheduling | ✗ | ✗                                                                | ✗ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|           Scenes | ✗ | ✗                                                                | ✗ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|     Effect Modes | ✓ | ✓                                                                | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |
|           Groups | ✓ | ✓                                                                | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓                                | ✓ | ✓                                | ✓                                |

### Links to docs

- [MiLight API "docs"](http://www.limitlessled.com/dev/)
- [LifX API docs](https://api.developer.lifx.com/)
- [Hue API docs](http://www.developers.meethue.com/philips-hue-api)

**Comment by eliotstocker** [comment]
great writeup! I have started work on the Light control Provider and hope to have a good framework ready soon :)

**Comment by eliotstocker** [comment]
Hi all, whilst the implementation is not complete i have done quite a lot on a provider model, its become part of the API so check it out here:
http://eliotstocker.github.io/Light-ControllerAPI/#provider-api

this should allow (once completed) for the community to create providers that allow all sorts of Lights etc to be controlled within this application

**Comment by dudeofawesome** [comment]
> All sorts of lights etc to be controlled

Hmmm… I should write a provider for my garage door. Brightness will translate to how open it is 😄

**Comment by eliotstocker** [comment]
lol

**Comment by dudeofawesome** [comment]
So if I understand this correctly, if I wanted to add support for [some other lightbulb here], as an end user, I'd have to download a separate app from the Play Store?

**Comment by eliotstocker** [comment]
that is correct, i will however create a dialog to make it really easy inside of light controller

**Comment by dudeofawesome** [comment]
Okay, that sounds good. We might want to specify a unique keyword authors should tag their apps with, that way searching in the Play Store is easy.

**Comment by eliotstocker** [comment]
definitely a good idea, i with we could just search by intent category or something, but it doesn't seem so. maybe could add something like LCProvider to every app title?

i more than happy to publish anyones Providers for them if they dont want to sign up for the dev console etc also.

**Comment by dudeofawesome** [comment]
I believe that adding a keyword would be sufficient.

**Comment by eliotstocker** [comment]
ok cool :) haven't looked into it much myself

**Comment by mrwhale** [comment]
Check out Juice SSH, they have a plugin system for their main app, and they seem to handle it pretty well. Seems the keywork should be enough. Even LightController. So "LightController Hue Plugin"?

But having a page within settings to show the available plugins would be handy too

![screenshot_20161018-034927](https://cloud.githubusercontent.com/assets/7307429/19446770/03c7e01e-94e6-11e6-8518-8bd4db00ef5a.png)
